[16.0][MIG] account_customer_wallet

carmenbianca opened the pull request:
## Description


## Odoo task (if applicable)

https://gestion.coopiteasy.be/web#id=10368&action=479&model=project.task&view_type=form&menu_id=536

## Checklist before approval

- [ ] Tests are present (or not needed).
- [ ] Credits/copyright have been changed correctly.
- [ ] Change log snippet is present.
- [ ] (If a new module) Moving this to OCA has been considered.

carmenbianca committed:
[ADD] account_customer_wallet

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: is_customer_wallet_journal field

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: Global settings

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Make the global journal option mandatory if the global toggle is enabled

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: Create customer_wallet_balance for res.partner

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Use account instead of journal

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Save customer wallet account on res.company

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: Demo account/journal data

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: Scaffolding + logic for balance computation

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: Test what happens when the wallet account is un-configured

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Write account_wallet to res.company instead of settings

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Use search instead of filter

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Mark failing test as expectedFailure for now

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] pos_customer_wallet: Recompute balance based on POS statement lines

This does not yet live-update in the POS. The window needs to be
reloaded to see the updated balance.

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REF] account_customer_wallet: Rename account.py to account_account.py

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Remove some TODO comments

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REF] account_customer_wallet: Use a related instead of a computed field

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REM] account_customer_wallet: Remove some fields and api.depends lines

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REF] account_customer_wallet: Re-implement wallet computation using partner._invoice_total algorithm

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] pos_customer_wallet: Tests

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REF] pos_customer_wallet: Re-implement wallet computation using partner._invoice_total algorithm

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[IMP] fully customize demo data on res_company, making the module ready to test, once installable with demo data

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REM] account_customer_wallet: Custom wallet toggle; depend on whether account is set instead

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[ADD] test to see if balance is commown between child and parent partner

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Search balance on all partners in family tree

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: README.rst

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

Commit:
[UPD] Update account_customer_wallet.pot

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: Search on customer_wallet_balance

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] account_customer_wallet: Change log entry

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REF] *_customer_wallet: Create partner.get_all_partners_in_family

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] *_customer_wallet: Change log entries

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

Commit:
[UPD] Update account_customer_wallet.pot

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
[UPD] changelog

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
account_customer_wallet 12.0.1.1.0

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
[UPD] README.rst

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] SCRLfs → SC

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
[UPD] changelog

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
[UPD] README.rst

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
account_customer_wallet 12.0.1.2.0

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] SCRL fs → SC

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[IMP] Prevent payment in back-office if wallet balance is negative

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[ADD] new field 'is_customer_wallet_product' on product model for product used to credit wallet accounts

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[REF] Add GRAP as coauthor of the wallet modules

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[i18n] update pot files and add fr translations

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[TR] update fr translation

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[REF] create dedicated function get_wallet_balance_account_move_line to compute wallet balance, based on account.move.line. (as for the module pos_customer_wallet)

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[REF] refactor _compute_customer_wallet_balance overload. Do not duplicate alogithm

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[REF] Do not affect temporary value to odoo items, saving time. 2104 ms -> 1260 ms

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[REF] account_customer_wallet : Do not call get_all_partners_in_family() if it's not required. Saving time : 19801 ms -> 5875 ms

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[translation] *_customer_wallet : update translation

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[REF] Move minimum_wallet_amount field from pos.config to account.journal, cortesy @carmenbianca

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[DOC] Add comment to explain why tthe computation of the partners of the families are splitted

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[REF] apply minimum wallet amount also in main account_customer_wallet

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

Commit:
[UPD] Update account_customer_wallet.pot

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
[UPD] README.rst

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[MIG] account_customer_wallet: Migration to 13.0

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[MIG] account_customer_wallet: Migration to 14.0

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[MIG] account_customer_wallet: Migration to 15.0

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[MIG] account_customer_wallet: Migration to 16.0

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Adjust xpath

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Repair all tests that aren't related to invoices

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: First steps to account.invoice compatibility

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca commented on account_customer_wallet/tests/common.py:
```diff
@@ -0,0 +1,124 @@
+    def _create_payment(self, invoice, amount=0, partner=None):
+        if partner is None:
+            partner = self.partner
+        register = (
+            self.env["account.payment.register"]
+            .with_context(
+                active_model="account.move",
+                active_ids=invoice.ids,
+            )
+            .create(
+                {
+                    "journal_id": self.customer_wallet_journal.id,
+                }
+            )
+        )
+        return register._create_payments()
```
So this kind of works, and triggers the balance check in `account_payment.py` correctly, but it does NOT create `account.move.line`s that affect the customer account wallet. I do not know which code to alter where to create an invoice OR a payment register that credits the wallet account.

legalsylvain replied:
I just tested.
In V16, create a payment, by default, create an account move line with the account defined globally on the company in the field "Outstanding Receipts Account".
1) I don't know why.
2) I don't know how to handle correctly the payment in V16.

legalsylvain replied:
If we set the wallet account on the journal on that field, it works.

![image](https://user-images.githubusercontent.com/3407482/233695669-d682e4bd-47f9-4b3f-9372-970c153b9a1f.png)

carmenbianca replied:
Thank you Sylvain for that insight. I changed the code to make this work. Commits [66b4a5a](https://github.com/coopiteasy/addons/pull/290/commits/66b4a5a73efcd956e93d30726942878a6a392cb0) and [286a6e2](https://github.com/coopiteasy/addons/pull/290/commits/286a6e2de5ee78a90750fd2e91a05baa8ce998ac) should both be interesting to you. Some of that work may be backported to v12, maybe possibly.

carmenbianca commented on account_customer_wallet/models/account_payment.py:
```diff
@@ -0,0 +1,47 @@
+        related="partner_id.customer_wallet_balance",
+    )
+
+    def action_post(self):
+        wallet_payments = self.filtered(
+            lambda x: x.journal_id.is_customer_wallet_journal
+        )
+
+        for payment in wallet_payments:
+            if (
+                payment.partner_id
+                and (payment.customer_wallet_balance - payment.amount)
+                < payment.journal_id.minimum_wallet_amount
+            ):
+                raise UserError(
```
In manual testing, this check works, BUT it doesn't test whether any of the move lines actually credit the customer wallet account.

carmenbianca reviewed (commented):
I mostly succeeded porting this, except one test.

@legalsylvain You're the author of the code I'm struggling to port. Do you have any insights into this? 'No' is a valid answer :)

carmenbianca committed:
[FIX] account_customer_wallet: Remove redundant name strings

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain committed:
[IMP] account_customer_wallet : make the module working with global partners

**before**

the computation is based on that account  ``customer_wallet_account_id =
fields.Many2one(related="company_id.customer_wallet_account_id")``

so, if you have a "global" partner, there is no company so no account,
and the wallet feature will not work.

**After**

The computation is based on ``wallet_account =
self.env.company.customer_wallet_account_id``. so it depends on the
context. It will work for global partners, (no company_id) the value
will depend on the current context. (that makes sense in my opinion). If
you are in a company A, you have a CoA A and you have a wallet A amount.
if you are in a company B, you have a coA B and you have a wallet B
amount.

carmenbianca committed:
[IMP] account_customer_wallet: Verify account.move is posted before including it in balance calculation

The reason this test is added (apart from being objectively more
correct) is that this work is necessary to get the account.payment stuff
right. _create_payments() in account.payment.register creates the
necessary account.move.lines _before_ we do our balance test in
action_post() in account.payment.

This (almost) completely refactored
get_wallet_balance_account_move_line() to no longer use the where_query
stuff. For some reason, when adding the extra state domain check, the
database query returned nothing. So logically, one opens the debugger
and does a manual search on the domain in that method, before the
database query is executed. This returned the expected result.
Subsequently, when continuing the debugger, the database query would
return the expected result...

I wasn't able to investigate the cause of that discrepancy, so I just
rewrote the method to do a plain search on the database. Is there a
reason we didn't do this before? I can find this discussion:
https://github.com/coopiteasy/addons/pull/222/files#r856062451 belonging
to this commit:
https://github.com/coopiteasy/addons/commit/851a2b8e12bc6f30eae4c9d75ab667e832adc291

But the only identifiable benefit is being able to query multiple
partners, which we already do.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] account_customer_wallet: Fix the account.payment test

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain commented:
@carmenbianca : could you take a time to take a look on this one ? https://github.com/coopiteasy/addons/pull/291

it is now conflicting, because of recent change, but I can rebase, if you think this PR is interesting. (IMO, yes ;-), but let me know).

carmenbianca commented on account_customer_wallet/models/account_payment.py:
```diff
@@ -25,6 +25,9 @@ def action_post(self):
         )
 
         for payment in wallet_payments:
+            # TODO: Figure out why/whether this statement is needed. At time of
+            # writing, the tests do not pass without it.
+            payment.partner_id._compute_customer_wallet_balance()
```
I'm unhappy about this.

legalsylvain replied:
That's weird.

legalsylvain replied:
Do you want I investigate a little on that topic ?

carmenbianca replied:
If you're happy to do so. I'm leaving this PR alone for a little bit; other priorities this week.

legalsylvain replied:
"Happy" is a bit of a strong word ;-), but if I can help port this module that GRAP uses...

carmenbianca commented on account_customer_wallet/tests/common.py:
```diff
@@ -79,6 +79,9 @@ def _create_move(self, debit=0, credit=0, partner=None):
             }
         )
+        move.action_post()
+        partner._compute_customer_wallet_balance()
```
I'm also unhappy about this.

carmenbianca commented on account_customer_wallet/tests/test_balance.py:
```diff
@@ -68,8 +68,9 @@ def test_no_wallet_account(self):
-
         self.assertFalse(self.company_id.is_enabled_customer_wallet)
+
+        self.partner._compute_customer_wallet_balance()
```
And this.

carmenbianca committed:
[IMP] account_customer_wallet: Add api.depends to customer_wallet_balance

This was missing before, and caused undocumented behaviour (+ having to
recompute in random places).

With some tests added, this should be solid now.

Co-Authored-By: Sylvain LE GAL <sylvain.legal@grap.coop>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[IMP] account_customer_wallet: Add dependency on context company to balance computation

Co-Authored-By: Sylvain LE GAL <sylvain.legal@grap.coop>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

legalsylvain commented:
nice !

carmenbianca commented:
/ocabot merge nobump
